### PR TITLE
scratch ipv6 base classes

### PR DESCRIPTION
**Pull request by aarc0305**
1. Create a DataPlaneLPMIPv6GPU class in dataplane_lpm_gpu.h/.cpp
2. Create a IPv6LPMFactory class in cuda_lpm_factory.h/.cpp. However, the AddLPMRule function has not been implemented yet.

### CODE REVIEW

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 1]
```diff
@@ -31,6 +31,14 @@ __global__ void InitLPMTable(IPv4RuleEntry *ipv4_tbl_24) {
   ipv4_tbl_24[idx].external_flag = false;
 }
 
+__global__ void InitIPv6LPMTable(IPv6RuleEntry *ipv6_tbl_24) {
+  int idx = blockIdx.x * blockDim.x + threadIdx.x;
+  ipv6_tbl_24[idx].next_hop = 254;
+  ipv6_tbl_24[idx].valid_flag = false;
+  ipv6_tbl_24[idx].depth = 0;
+  ipv6_tbl_24[idx].external_flag = false;
+}
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu]: Should still initialize `tbl8_ptr` in `nullptr` or `NULL`, 
```cuda
ipv6_tbl_24[idx].tbl8_ptr = nullptr;
```

Comment (aarc0305) [reply gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu]: ok, I would add it.

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.h [hunk 2]
```diff
@@ -20,6 +20,14 @@ struct IPv4RuleEntry {
...
 };
 
+struct IPv6RuleEntry {
+  uint8_t next_hop;
+  bool valid_flag;
+  // The external flag will be used when the depth exceed 24 bits.
+  bool external_flag;
+  uint8_t depth;
+  IPv6RuleEntry *tbl8_ptr;
+};
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.h:30]: add a new line here.

File: gpuflow/dataplane/gpu/dataplane_lpm_gpu.h [hunk 3]
```diff
@@ -23,6 +23,17 @@ class DataPlaneLPMIPv4GPU {
...
+ public:
+  DataPlaneLPMIPv6GPU();
+
+  //Delegate pointer on this class.
+  IPv6RuleEntry *IPv6TBL24;
+ 
+ private:
+  cu::IPv6LPMFactory ipv6_lpm_factory;
+  int CreateLPMTable();
+};
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/dataplane_lpm_gpu.h:36]: add a new line here, too.

**Comment by tz70s** [comment]
Due to the review-modify cycle, I think it's better co-work on this branch in my repo.

**Comment by tz70s** [comment]
ref to #13

File: gpuflow/dataplane/gpu/dataplane_lpm_gpu.cpp [hunk 4]
```diff
@@ -35,5 +35,18 @@ int DataPlaneLPMIPv4GPU::CreateLPMTable() {
...
+  }  
+}
 
-} // namespace gpuflow
\ No newline at end of file
+int DataPlaneLPMIPv6GPU::CreateLPMTable() {
+  if(ipv6_lpm_factory.CreateLPMTable() < 0) {
+    std::cerr << "Create lpm table error from ipv6 lpm factory";  
+  } else {
+    std::cout << "Create lpm table from ipv6 lpm factory successfully." << std::endl;
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/dataplane_lpm_gpu.cpp]: Pass the delegate pointer from factory class into this class, as ipv4 did.

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 5]
```diff
@@ -66,5 +86,82 @@ int IPv4LPMFactory::AddLPMRule(uint32_t ipv4_address, uint8_t depth, uint8_t nex
...
+  int idx = threadIdx.x;
+  ipv6_tbl_24[start + idx].next_hop = next_hop;
+  ipv6_tbl_24[start + idx].valid_flag = true;
+  // TODO
+  // Currently, we only handle with the case of depth <= 24 
+  // The case of depth > 24 would use the tbl8 table.
+  ipv6_tbl_24[start + idx].external_flag = false;
+  ipv6_tbl_24[start + idx].depth = depth;
+  ipv6_tbl_24[start + idx].tbl8_ptr = nullptr;
+}
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu]: At here, we'll need to check if it is a longest rule.
```c++
if (ipv6_tbl_24[start + idx].valid_flag && (ipv6_tbl_24[start + idx].depth > depth)) {
  // There's an existed rule and longer, abort this update
} else {
  // Add the new rule
  // ...
}
```

File: gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu [hunk 6]
```diff
@@ -66,5 +86,82 @@ int IPv4LPMFactory::AddLPMRule(uint32_t ipv4_address, uint8_t depth, uint8_t nex
...
+    for(unsigned long int i = 0; i < (24 - depth); i++) {
+      distance *= 2;
+    }
+
+    // Calculate the start and end.
+    unsigned long int right_shift = ipv6_address[1] >> (16 - depth);
+    start = ipv6_address[0] << 16 | right_shift << (24 - depth);
+    end = ipv6_address[0] << 16 | right_shift << (24 - depth) + distance;
+
+  } else if (1 <= depth < 8) {
```
Comment (tz70s) [inline gpuflow/dataplane/gpu/cuda/cuda_lpm_factory.cu]: The depth of `0` can be also added.
i.e.
```bash
sudo route add 0.0.0.0 dev dtap0
```

**Review by tz70s — CHANGES_REQUESTED** [review]
Great work! Add the two needed changes, then we can merge and test it first.

**Review by tz70s — APPROVED** [review]

**Pull request merged**